### ISSUE TITLE
Mechanism to hide lasers

### ISSUE THREAD

**Issue by shiftyscales** [Frooxius/NeosPublic#503]
Creating this issue as a follow-up to #485, having a mechanism of some form to disable/block laser visuals could be handy for certain use-cases.

**Comment by Frooxius** [comment]
Can you elaborate the use-cases? There are several possibilities to hide the lasers, but understanding when it's needed would help tailor a good solution.

**Comment by shiftyscales** [comment]
Sure. Among the feedback I've heard:
- GearBell wishes to be ableto remotely puppeteer avatars/objects, and use it for 'shenanigans' (moving objects around someone discretely in a social setting)
- Zane/DeliriousJax wants to be more 'immersed' like the previous physical mode (hide toolshelves and disable lasers)
- I've heard another case where a user wished to use their trigger for controlling their blendshapes without activating their lasers

**Comment by DeliriousJax** [comment]
Custom facial expression gestures not activating lasers, as well as puzzle maps that will have items out of reach and you need to get to said items/have a partner be able to reach the items not using lasers being another reason to add on.

**Comment by DeliriousJax** [comment]
Based on when I personally toggled physical mode and now scale locking/reset scale, disable lasers and scale lock/reset would feel very natural together/next to one another in some fashion.

**Comment by Frooxius** [comment]
I think there are actually several things that should be added to cover for different scenarios, because it sounds like each of those examples has different requirements. 

E.g. remote puppeteering - in this case GearBell needs to see her own lasers so she can actually remote interact with the items, but they should be hidden from other users.

For the immersion, it sounds like the remote grabbing interaction (or any remote interaction) should be disabled altogether, so it's not just about hiding the lasers from other users, but not being able to bring them up at all, even for the user themselves.

For the gestures, question is if that could be covered by disabling them for yourself. It might be annoying to constantly turn them on and off again if you want to make expressions in the meanwhile. Or is the common scenario that you want to keep them off for a longer time (meaning you won't be able to interact at all either)?

It might be worth considering some heuristic on when to not toggle them in that case, but we also need to make it so it's not confusing for new users.

**Comment by DeliriousJax** [comment]
Based on where/how often new users used the physical mode before when I would introduce them to neos, it's often less technical use case and more social appearance/self sensation. It never came up as an issue of confusion to use physical mode once they knew where it was (as far as I ever saw) and I observed it toggled on often.

Puzzle maps aside, some of the projects a lot of people want to begin working on and releasing, including myself, a whole other community I've been slowly introduced into Neos, and even gearbell has mentioned a desire to use neos for is story and roleplay environments. The key to immersion being the ability for each user to hide/toggle their toole and shelves at need based on whether they are in a circle and talking to people and using many hand gestures/custom control expressions or interacting with map objects.

**Comment by Frooxius** [comment]
What was the main reason they used physical mode before? From what I heard it was to prevent scaling, but that's now handled by a separate lock, which will be extended later on.

Same goes for other behaviors as well previously covered by physical mode, like hiding lasers and so on. The new lasers were designed so they don't keep coming up on their own, but still let users interact with things in casual setting. 

Because the system is being redesigned, I need to understand what were the reasons and requirements of the old systems, so we can design proper new solutions to cover those.

For building games, stories and so on, it might be better to have the ability to disable those interactions for the user in the particular world/context to prevent them from cheating or accidentally breaking the immersion. I want to push more controls like that, giving the creators more control then before on how are such worlds setup, rather than requiring on specific quirks of the base interaction system.

**Comment by ProbablePrime** [comment]
I think the issue here is the collaborative design and management of such experiences. Consider for a second the Meta Movie project. In this scenario some actors or staff within the experience may need to use lasers or tool shelves within a world. But the player probably shouldn't be using them.

In this case a fine grain user based on/off for lasers, Tool Shelves and other auxiliary UI is needed. Its not as simple as just turning it off for the entire world as some users may need it in the world.

**Comment by Frooxius** [comment]
I'm sorry, I didn't mean to imply that this would be for all users per world. As with anything, you could configure/script which users would have the lasers forced off and which ones not. This is just one proposed mechanism as well, not mutually exclusive with others. I don't think there's going to be a single fit-all approach.

I'm just trying to distill the requirements and different systems/settings to accommodate those.

E.g. the ability to hide your own lasers is separate from the world settings. E.g. if the world disables them for you, you won't be able to bring them on. This can prevent cheating and other problems if the lasers would break your game.

If the world allows lasers (for the user specifically), then the ability to toggle them on and off would be up to you.

Neither of those cases would cover GearBell's scenario though, because she needs the lasers on, but just have not other users see them. 

That's why I'm trying to look what would be the best way to structure the system to allow fine-grained control to cover all the use-cases, while at the same time making the system simple for new users and sustainable long term.

**Comment by ProbablePrime** [comment]
Understood:

> E.g. the ability to hide your own lasers is separate from the world settings. E.g. if the world disables them for you, you won't be able to bring them on. This can prevent cheating and other problems if the lasers would break your game.
> 
> If the world allows lasers (for the user specifically), then the ability to toggle them on and off would be up to you.

This description covers the use cases im interested in. Thank you for thinking it through!

**Comment by DeliriousJax** [comment]
That definitely helps clarify a bit more, thank you.

**Comment by Frooxius** [comment]
In 2020.5.8.141 I've implemented the lasers activating on partial trigger press and deactivating immediately when you let go. This allows you to activate the laser, grab something and let go, which will make the laser go away (unless you do a full trigger click).

This could cover GearBell's scenario for "Shenanigans", but I'm not sure how important is the initial visibility of the laser in the case. Can you let me know how this feels?

This could be interesting mechanism though. But we'll see if it leads to any problems and confusion though, but it's worth a try I think.

For other scenarios I still have to design some solutions.

**Comment by ProbablePrime** [comment]
Just adding that I'd like the ability to turn off my toolshelves in a native manner somehow. I was going to make a separate issue but saw mention of them here so wanted to comment here instead. I'm aware that there are LogiX related stuff i can do to hide them but I would rather have something native. 

Additionally an edit mode that doesn't have the Grip Pose references would be good. I sometimes use edit mode for... editing and dont need my grip pose references to continue editing the world.

So in summary:
- Native way to hide toolshelves
- Native way to hide Grip Pose reference hands while in edit mode.

**Comment by DeliriousJax** [comment]
Information regarding situations in which more interaction options would be useful:

Recently I launched (this far un-announced) a roleplay map that will be hosted on my and Zane's machines moving forward.

While hosting events or placing items to decorate the map and plot for what the players will interact with, we would like a quick and easy way to toggle our lasers and tool shelves on and off to quickly swap between our in character state and out of character state. That way we dont have to continuously respawn the tools we need and delete them to keep from cluttering our tool shelves and breaking immersion. 

On top of this it will let us more easily use custom emotes and scripted non laser using sliders and gadgets without activating our lasers and risking accidentally moving far/undesired map props/decoration.

It will also allow our players to quickly take control of items they spawn from their inventory to move and rescale, treat their tool shelves like inventories/pockets if needed and for the same immersing, non intrusive reasons listed above.

Another thing. When not equipping a tool, it would be nice to be able to grab a distant object and move it around for in character reasons -without- our lasers being visible to others at all, allowing us to puppeteer objects during events without the blatant "puppet strings" of the lasers. This goes in line with one of gearbell's requests for hidden lasers. Only such tools as the dev tool should b3 visible to other players for technical and instructive "laser pointer" use when teaching new players.

Hopefully this helps present clearer views on just a few of the (many unmentioned) possibilities and reasons for the native laser and tool shelf toggle/control requests.

**Comment by DeliriousJax** [comment]
Feedback on the partial laser press. Oculus Rift S. Can't really feel it and it seems relatively difficult to find the pressure to get it. It also does not work for the "Shenanigans" mentioned as the lasers are visible to other players which is the issue when trying to sneakily move props around without people knowing who it is.

**Comment by Hayden-da-Serg** [comment]
Responding to Jax, I didn't think I was crazy, the strength required to toggle the laser is the same strength required for me to send the TriggerClick event so it feels the exact same to me as it did before.
I'm personally running on an Oculus Quest playing through VIrtual Desktop (which might as well be a wireless Rift S running through SteamVR)

**Comment by shiftyscales** [comment]
@DeliriousJax @Hayden-da-Serg- this is unfortunately due to the fact the "click" on Oculus controllers is extremely shallow. Around 13% or so through the full trigger range (this is slightly higher on SteamVR).

Once the setting are overhauled, it will be possible to expose the threshold value for the soft-pull trigger- but for now, precaution was taken so as not to allow the laser to accidentally activate- the threshold for it is incredibly low as it is already- the issue is unfortunately in how Oculus has mapped the trigger. It's not something we can change.

It is possible to soft-pull both through Oculus Native, and SteamVR, I made sure to test that (it literally was impossible to do so prior to the recent tweaks), but fine-grained control can be made available after the settings are overhauled.

**Comment by Frooxius** [comment]
Yeah as @shiftyscales said, we unfortunately discovered that Oculus considers very shallow press a click (around 18 % with native, 25 % with SteamVR), while on Vive/Index it's 100 % press, which gives very little room for the soft activation before the full click happens.

One thing I could do is modify the input driver so full click on Oculus is 100 % as well, ignoring the click coming from the SDK, but this will affect everything, meaning you'll need a full press when clicking the UI.

I was considering option of doing this for lasers as well, but this presents a problem as well, as the actual UI click would happen before the lasers are fully activated, which would mess the system up, as you'd now be clicking with non-activated laser if that makes sense. 

@DeliriousJax Just to clarify the point, is the problem with the shenanigans that the lasers are seen initially when grabbing, even if they're invisible afterwards?

**Comment by DeliriousJax** [comment]
@Frooxius The lasers remain visible to other people. If we ourselves can see our own lasers while holding something, thats fine. But we want our lasers invisible to other people.

**Comment by Frooxius** [comment]
@DeliriousJax Are we talking about the new mechanism which lets you hide the lasers while grabbing something? In this case the lasers are not visible even to you, so they should be hidden from other players.

They're only visible briefly when you initiate the grab, but not afterwards after you let go of the trigger. It's probably trickier to execute on Oculus though, as you need to just soft-pull activate them, grab the item and then let go of the trigger. This makes the laser disappear while the item is still held.

**Comment by DeliriousJax** [comment]
In that case, such a mechanism is definitely not going to work in this form of implementation. It may be worth adding to the list of things to consider in the play modes presets we've discussed in the past. Which based on how that idea has evolved thus far, I may write a git post for of it's own.

**Comment by Frooxius** [comment]
Can you explain why a bit please? Is it the brief initial visibility of the laser that prevents the shenanigans? Or the difficulty triggering on Oculus? Or something else? I can't do much tweaking or finding alternate solutions if I can't understand which part is causing the problems.

That said, I am still planning to add other parallel mechanisms - ability to disable lasers by the world owner / LogiX, to cover for game items and worlds.

But I'm trying to see if this activation mechanism could work well for the "shenanigans" part.

**Comment by DeliriousJax** [comment]
Based on the majority of what we discuss in game. -Native- solutions via toggles and settings and play modes is what most of us collectively want for sure, definitely. The logix rout has already been explored by probableprime and tested on a few people and it's already fading from interest via that method. 

As for faded lasers. I saw it was implemented but its so fine and difficult to trigger, none of us even knew how to get it working. Might be oculus, might be its present settings, but either way, that implementation is very temporary/bandaid solution based on what we were looking for.

**Comment by Frooxius** [comment]
Again, can you please explain why? I know what you want, but I need to make sure it fits well with the systems in Neos and so it's designed properly, rather than a ad hoc solution that will cause problems in the future and possible breaking of content again.

That's something I want to prevent with this system by properly designing it well from scratch, but I need to understand what you need and why certain solutions don't seem to work or why you wouldn't want them.

Why do you feel that LogiX route is fading from interest? I don't think this is quite representative, since currently there is no proper LogiX mechanism, so your options here are highly limited.

If it's difficult to trigger, we can look into that. How do you feel about the adjustment of the trigger threshold on Oculus? I don't see why exactly it would have to be temporary/bandaid solution, the idea of it is that it's a neat gesture that covers one of the use-cases, we can still have others.

**Comment by DeliriousJax** [comment]
I mean I'm trying to give use case scenarios, but I'm running out of ways to explain it. Lets say I have a horde of spiders chasing a posse of sessioneers in a setting I've created for them and i need them to feel a little worried, trying to quickly move the spiders around cause a little chaos is very difficult to do while im required to constantly be gentle and figure out a partial press to keep my lasers hidden from the players.

Having logix or something dictate whether our lasers are shown is an undesired option because if we have to go through our avatar and remove gesture controlled objects just to interact with the map, or load a different avatar all together to make some edits and changes it is several more steps then to do rather than just opening a quick menu and saying enable/disable lasers/tool shelfs.

Simply put, as hacky as physical mode was. We love what it let us do. As simple and seemingly needless as it may look from a technical/programming perspective, it was an important integration for us from a community and social standpoint that allowed us far broader visions.

I'm honestly not sure what you mean by specific use cases if the above scenario with the spiders and other examples I've written up in previous replies aren't forms of said examples you're asking for.

**Comment by GearBell** [comment]
in the same location as the laser settings in Settings, itd be nice to have a checkbox for Toggle Laser Global Visability. That way we can just turn it on/off as a whole

**Comment by Frooxius** [comment]
Thanks for the details! In 2020.5.18.573 I've added quick option to the context menu that lets you turn your lasers on and off on each and. 

This should help with some cases like the Jax' flying hat, the wing flap flying mechanic and so on. It doesn't quite cover the "shenanigans" though, that needs a different control (whether lasers are visible to other users or not) and doesn't control the games either (forcing the lasers off), those are going to be separate things.

I plan on extending this into a full facet, which will have the individual laser controls. That should hopefully come relatively soon now.

**Comment by Ruikio** [comment]
Digging this back up as having an option to use my laser but not show it to other people would be useful.

Seeing the 'Needs more information' is very odd.

Let me and others use our lasers but not show it.

An example:
Dungeon Masters (DM) are invisible to the players and setting up scenes and objects. A key item was moved due to some moment in the roleplay on accident. The DM can laser pick up the item and put it back. The players don't have to get pulled out of immersion by seeing a giant laser.

**Comment by DeliriousJax** [comment]
@Ruikio Gearbell had some similar input, yes. I think it was placed on a shelf to complete when workspaces became a thing later.

**Comment by Ruikio** [comment]
@DeliriousJax I mean.. simply making it local via world settings or something seems like a simple solution for this. Not sure why it'd need to be packaged with a bigger feature.

**Comment by DeliriousJax** [comment]
@Ruikio my best guess is that it's one of many features desired to be in work spaces. When Froox and I initially came up with the idea of work spaces, it included a lot of these types of toggles and feature organizers.

**Comment by Ruikio** [comment]
> @Ruikio my best guess is that it's one of many features desired to be in work spaces. When Froox and I initially came up with the idea of work spaces, it included a lot of these types of toggles and feature organizers.

Fair. Well hopefully this topic gets looked at for the feature of hiding the user's lasers so other's can't see it as an option to the player or the world permissions and this will be one less thing to worry about via that topic.

I Rukio esquire the third am pro hide lasers. My immersion must be untouched!
